src/common/io/filter/group.c

```diff
@@ -45,29 +78,100 @@ Add a filter
+static IoFilterData *
+ioFilterGroupGet(IoFilterGroup *this, unsigned int filterIdx)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(IO_FILTER_GROUP, this);
+        FUNCTION_TEST_PARAM(UINT, filterIdx);
+
+        FUNCTION_TEST_ASSERT(this != NULL);
+    FUNCTION_TEST_END();
+
+    FUNCTION_TEST_RESULT(IO_FILTER_DATA, (IoFilterData *)lstGet(this->filterList, filterIdx));
+}
+
+/***********************************************************************************************************************************
+Open filter group
+
+Setup the filter group and allocate any required buffers.
+***********************************************************************************************************************************/
+void
+ioFilterGroupOpen(IoFilterGroup *this)
+{
+    FUNCTION_DEBUG_BEGIN(logLevelTrace);
+        FUNCTION_DEBUG_PARAM(IO_FILTER_GROUP, this);
+
+        FUNCTION_DEBUG_ASSERT(this != NULL);
+    FUNCTION_DEBUG_END();
+
+    MEM_CONTEXT_BEGIN(this->memContext)
     {
-        MEM_CONTEXT_BEGIN(this->memContext)
+        // If the last filter is not an output filter then add a filter to buffer/copy data.  Input filters won't copy to an output
+        // buffer so we need some way to get the data to the output buffer.
+        if (lstSize(this->filterList) == 0 || !ioFilterOutput((ioFilterGroupGet(this, lstSize(this->filterList) - 1))->filter))
+            ioFilterGroupAdd(this, ioBufferFilter(ioBufferNew()));
+
+        // Create filter input/output buffers.  Input filters do not get an output buffer since they don't produce output.
+        Buffer *lastOutputBuffer = NULL;
+
+        for (unsigned int filterIdx = 0; filterIdx < lstSize(this->filterList); filterIdx++)
         {
-            this->filterList = lstNew(sizeof(IoFilter *));
+            IoFilterData *filterData = ioFilterGroupGet(this, filterIdx);
+
+            // Assign the last output buffer to the input.  At first there won't be an input filter because it will be passed into
+            // the process function as an input.
+            if (lastOutputBuffer != NULL)
+            {
+                filterData->input = lastOutputBuffer;
+                filterData->inputLocal = lastOutputBuffer;
+            }
+
+            // Is this an output filter?
+            if (ioFilterOutput(filterData->filter))
+            {
+                // If this is the first input buffer found, store the index
+                if (lastOutputBuffer == NULL)
+                    this->firstOutputFilter = filterIdx;
+
+                // If this is not the last output filter then create a new output buffer for it.  The output buffer for the last
+                // filter will be provided to the process function.
+                if (filterIdx < lstSize(this->filterList) - 1)
+                {
+                    lastOutputBuffer = bufNew(ioBufferSize());
```

Comment:
I know storage.c is not in this review, but I don't understand why there are 2 calls to ioBufferSize with parameters (ioBufferSize(source) and ioBufferSize(file)). I cannot find a corresponding function that takes a parameter. Can you help me understand?

Reply:
This is because we did not have `-Wstrict-prototypes` enabled and I did a search and replace without properly checking my work.  A function declared `()` instead of `(void)` defaults to the old K&R style where params did not need to be explicitly defined.  I thought `-std=c99` forbid this style but apparently not.

I have enabled `-Wstrict-prototypes` and explicitly declared all functions as (void) where needed: 429a356e3323acf99f3d.  The examples you found were the only places where params were being passed in the old K&R way.